src/CommandFactory.ts

```diff
@@ -150,84 +163,50 @@ export class CommandFactory {
-            case Enums.Key.A:
-                return this.createEndAppendAction();
-            case Enums.Key.o:
-                return this.createInsertNewLineAction(false);
-            case Enums.Key.O:
-                return this.createInsertNewLineAction(true);
-            case Enums.Key.x:
-                return this.createCharactorDeleteAction(count);
-            case Enums.Key.X:
-                return this.createBeforeCharactorDeleteAction(count);
-            case Enums.Key.s:
-                return this.createCharactorDeleteInsertAction(count);
-            case Enums.Key.S:
-                return this.createLineDeleteInsertAction(count);
-            case Enums.Key.D:
-                return this.createDeleteToEndAction();
-            case Enums.Key.Y:
-                return this.createYancToEndAction();
-            case Enums.Key.C:
-                return this.createDeleteInsertToEndAction();
-            case Enums.Key.p:
-                return this.createPasteAction(false, count);
-            case Enums.Key.P:
-                return this.createPasteAction(true, count);    
-            // TODO
-            default:
-                return new PanicAction();
-        }
+        var char = Utils.KeyToChar(key);
+
+        var method: string = this.findMethodInKeybinding(key, [
+            keybindings.normalMode.singleAction,
+            keybindings.normalMode.textObjectOrSingleAction
+        ]);
+
+        return this[method]();
+    }
+
+    private insertAction(): IAction {
+        return new InsertAction();
+    }
+
+    private firstInsertAction(): IAction {
+        return new FirstInsertAction();
     }
 
     private createMotion(key: Enums.Key, count: number): IMotion {
         var m: IMotion;
-        switch (key) {
-            case Enums.Key.l:
-                m = new RightMotion();
-                break;
-            case Enums.Key.h:
-                m = new LeftMotion();
-                break;
-            case Enums.Key.j:
-                m = new DownMotion();
-                break;
-            case Enums.Key.k:
-                m = new UpMotion();
-                break;
-            case Enums.Key.n0:
-                m = new FirstMotion();
-                break;
-            case Enums.Key.Doller:
-                m = new EndMotion();
-                break;
-            case Enums.Key.w:
-                m = new ForwardWordMotion();
-                break;
-            case Enums.Key.b:
-                var bm = new ForwardWordMotion();
-                bm.SetBack();
-                m = bm;
-                break;
-            default:    
-                throw new Error("Panic!");
+        var char = Utils.KeyToChar(key);
+        var motion = this.findMethodInKeybinding(key, [
+            keybindings.normalMode.motion,
+            keybindings.normalMode.zero
+        ]);
+        var hasChain = motion.indexOf('.') !== -1;
+
+        if (hasChain) {
```

Comment:
I'm currently using this to allow for reverse calls on motions. (aka .SetBack). Not sure how much I'm happy with it yet.